Implemented Grostl

gsingh93 opened the pull request:
This PR contains an implementation of [Grostl](http://www.groestl.info/). I mainly used [this document](http://www.groestl.info/Groestl.pdf) as a reference for implementation. [This page](http://www.groestl.info/implementations.html) contains a guide for different optimizations one can do in an implementation, as well as many optimized reference implementations, but I chose to implement none of these. The goal of my implementation is that it should be correct, readable, and well tested. This is because none of the other implementations are readable or well-tested, which made using them as a reference difficult. After this is merged, we can work on optimizations, as we'd have a nice reference implementation to test against.

One way Grostl differs from most other hash functions is that it allows you to specify a variable digest size between 1 and 64 bytes. However the block size is different between digests of size 1 to 32 and 33 to 64. In order to prevent people from using the wrong block size, I've added `GrostlSmall` and `GrostlBig` which should prevent you from making this mistake when you specify your custom `OutputSize`. I've also provided type aliases for the common digest sizes, 224, 256, 384, and 512.

While the code is done, two minor things I plan on adding are more comments/documentation, more/better tests (especially around edge cases like padding), and cleaning up some of the assertions. It should be fine for you to start reviewing now, as these are minor changes.

gsingh93 committed:
Added stub Grostl implementation and tests

gsingh93 committed:
Implemented compress

gsingh93 committed:
Implement input and result Digest functions

gsingh93 committed:
Initialize Grostl state to correct value

gsingh93 committed:
Implemented Grostl padding function

gsingh93 committed:
Ensure max OutputSize for GrostlBig is 512 bits

gsingh93 committed:
Moved Grostl implementation to grostl.rs

gsingh93 committed:
Implemented Grostl finalize

gsingh93 committed:
Move all bounds into where clauses

gsingh93 committed:
Implemented Grostl P, Q, and sub_bytes

gsingh93 committed:
Implemented a generic Matrix type to use in the Grostl compression function

gsingh93 committed:
grostl: Added shift_bytes, test_shift_bytes, fixed some bugs, moved Matrix to its own mod

gsingh93 committed:
grostl: implemented add_round_constant

gsingh93 committed:
grostl: implemented mix_bytes

gsingh93 committed:
grostl: convert BlockSize and OutputSize from bits to bytes

gsingh93 committed:
grostl: fixed bug where matrix row and column were flipped

gsingh93 committed:
grostl: handle overflow in Matrix::mul_array

gsingh93 committed:
grostl: fixed padding bugs

gsingh93 committed:
grostl: implemented more unit tests and fixed various bugs

gsingh93 committed:
grostl: use digest buffer to be able to handle partial input

gsingh93 committed:
grostl: handle padding if less than 8 bytes are remaining, fix tests

gsingh93 committed:
grostl: don't depend on std, add crate doc block, define more Grostl aliases

gsingh93 commented on grostl/src/lib.rs:
```diff
@@ -0,0 +1,66 @@
+pub type Grostl224 = GrostlSmall<U28>;
+pub type Grostl256 = GrostlSmall<U32>;
+pub type Grostl384 = GrostlSmall<U48>;
+pub type Grostl512 = GrostlSmall<U64>;
```
Oops. The fact that `GrostlSmall` compiles with a size greater than `U32` is a problem...

newpavlov replied:
Thank you for your work!

Unfortunately it seems that according to E0122 "trait bounds are not (yet) enforced in type definitions" and because of the [bug](https://github.com/rust-lang/rust/issues/39122) I've missed it. (also you have typo in `GrostlSmall` which got through because of that, you needed to use `Cmp<U33>`)

So you'll need to write structs explicitly like this:
 ```Rust
#[derive(Default)]
pub struct GrostlSmall<OutputSize>
    where OutputSize: ArrayLength<u8> + Cmp<U33>,
          Compare<OutputSize, U33>: Same<Less>
{
    grostl: Grostl<OutputSize, U64>,
}

impl<OutputSize> Digest for GrostlSmall<OutputSize> {
    type OutputSize = OutputSize;
    type BlockSize = U64;

    fn input(&mut self, input: &[u8]) {
        self.grostl.process(input);
    }

    fn result(mut self) -> GenericArray<u8, Self::OutputSize> {
        self.grostl.finalize()
    }
}
```

It's probably not worth to implement `Digest` for `Grostl` in this situation.

I will review your code shortly.

newpavlov commented on grostl/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "grostl"
+version = "0.3.0"
```
It's better to start from 0.1.0 as we don't require version synchronization across crates.

newpavlov commented on grostl/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+authors = ["The Rust-Crypto Project Developers"]
+license = "MIT/Apache-2.0"
+description = "Grostl hash function"
+documentation = "https://docs.rs/md5"
```
Wrong link to docs.

newpavlov commented on grostl/src/grostl.rs:
```diff
@@ -0,0 +1,492 @@
+use generic_array::typenum::{Quot, U8};
+use matrix::Matrix;
+
+const SBOX: [u8; 256] = [
```
I think it's better to move constants to the separate module, in other crates I've used `consts.rs` for this purpose.

newpavlov commented on grostl/src/grostl.rs:
```diff
@@ -0,0 +1,492 @@
+const SHIFTS_Q_WIDE: [u8; 8] = [1, 3, 5, 11, 0, 2, 4, 6];
+
+pub struct Grostl<OutputSize, BlockSize: ArrayLength<u8>>
+    where BlockSize::ArrayType: Copy
```
Add `#[derive(Copy, Clone)]` to `Grostl` and `GrostlState`. You'll need to add several `BlockSize::ArrayType: Copy` trait bounds for that to work.

newpavlov commented on grostl/src/grostl.rs:
```diff
@@ -0,0 +1,492 @@
+        let block_bytes = BlockSize::to_usize();
+        if block_bytes == 64 {
+            false
+        } else {
```
Maybe use match statement with `unreachable!` macro?

newpavlov commented on grostl/src/grostl.rs:
```diff
@@ -0,0 +1,492 @@
+#[cfg(test)]
+mod test {
+    use super::{xor_generic_array, C_P, C_Q, Grostl, SHIFTS_P};
+    use generic_array::typenum::{U32, U64};
```
I think it's better to move this test module to the `tests.rs` and use:
```Rust
#[cfg(test)]
mod tests
```
In the `lib.rs`.

newpavlov replied:
On the second thought it will require to make tested methods public, they still will not be accessible to the crate user as `grostl` field will be private, but perhaps it's better to leave it as is. Your choice.

newpavlov commented on grostl/src/grostl.rs:
```diff
@@ -0,0 +1,492 @@
+
+struct GrostlState<OutputSize, BlockSize: ArrayLength<u8>> {
+    state: GenericArray<u8, BlockSize>,
+    rounds: u8,
```
You also can impl `Default` for `GrostlState`. This will allow you to simplify code and tests a bit and you will be able to simply use `#[derive(Default)]` on `Grostl`.

newpavlov commented on grostl/tests/lib.rs:
```diff
@@ -0,0 +1,36 @@
+extern crate grostl;
+
+use crypto_tests::hash::{Test, main_test};
+use generic_array::typenum::{U28, U32, U48, U64};
```
You can use type aliases instead of using `GrostlSmall` and `GrostlBig` directly. In my opinion it will look a bit nicer.

newpavlov reviewed (changes requested):
It's just a cursory review. I haven't looked deep into implementation yet, but it looks good.

Though formatting differs from used in other crates. But it's my fault, as I haven't got time to make rustfmt config. (personally I am not fond of multiline approach of rustfmt)

Also, please add LICENSE files as in other crates.

newpavlov commented:
By the way, correct name for Grøstl is Groestl, not Grostl. So it's better to rename crate. :)

Adding "grostl" to kewords should help with searchability.

gsingh93 committed:
grostl: cleanup based on review comments

gsingh93 commented:
Ok, I believe I've addressed all of your comments. The crate and all structs should be renamed, the Small/Big issue should be fixed, and the other minor code changes were made. I decided to leave the tests in the same mod, as those were unit tests which need access to the private functions. The tests in `tests` are integration tests and only use the public API.

I still need to add some tests to ensure the padding is correct, I haven't convinced myself that it is. I'm also planning on adding a benchmark, as I see there's one in the utils crate.

Also, what names do I put for the copyright in the LICENSE file? My own?

newpavlov commented on groestl/LICENSE-MIT:
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2006-2009 Graydon Hoare
+Copyright (c) 2009-2013 Mozilla Foundation
+Copyright (c) 2016 Artyom Pavlov
```
It's your work, so it's better to attribute yourself.

newpavlov commented on groestl/src/groestl.rs:
```diff
@@ -0,0 +1,438 @@
+    fn result(mut self) -> GenericArray<u8, Self::OutputSize> {
+        self.buffer.next(1)[0] = 128;
+        if self.buffer.remaining() >= 8 {
+            let block_bytes = self.block_bytes();
```
You can use `standard_padding` method from `digest-buffer` to reduce amount of code.

newpavlov commented on groestl/src/groestl.rs:
```diff
@@ -0,0 +1,438 @@
+
+impl<OutputSize, BlockSize> Digest for Groestl<OutputSize, BlockSize>
+    where OutputSize: ArrayLength<u8>,
+          BlockSize: ArrayLength<u8> + Div<U8>,
```
Not sure if it's worth to implement `Digest` for `Groestl`. You can move `input` and `result` to struct's `impl`. It'll reduce amount of code a bit.

newpavlov commented on groestl/src/groestl.rs:
```diff
@@ -0,0 +1,438 @@
+    #[test]
+    fn test_shift_bytes() {
+        let g: Groestl<U32, U64> = Groestl::default();
+        let s = g.state;
```
You do not use `g` anywhere after that. Maybe just write `let s = GroestlState::<U32, U64>::default();`? The same goes for other tests as well.

newpavlov commented on groestl/src/groestl.rs:
```diff
@@ -0,0 +1,438 @@
+            48, 57, 2, 11, 20, 29, 38, 47,
+            56, 1, 10, 19, 28, 37, 46, 55,
+        ];
+        assert_eq!(&*block, &expected[..]);
```
You can write it as `assert_eq!(&block[..], &expected[..]);`, which will be a bit clearer in my opinion. The same goes for other tests. I'll probably create later PR for `generic-array` with implementation of `PartialEq<[T]>`.

newpavlov commented on groestl/src/lib.rs:
```diff
@@ -0,0 +1,128 @@
+//! let mut hasher = GroestlBig::<U52>::new();
+//! hasher.input(b"my message");
+//! let result = hasher.result();
+//!}
```
Unnecessary `}`.

newpavlov commented on groestl/src/lib.rs:
````diff
@@ -0,0 +1,128 @@
+//! ```
+//!
+//! ```rust,ignore
+//! use groestl::{Digest, Groestl256};
````
Should be `GroestlSmall` and `GroestlBig`, not `Groestl256`.

newpavlov commented on groestl/src/lib.rs:
```diff
@@ -0,0 +1,128 @@
+mod matrix;
+
+#[derive(Default)]
+pub struct GroestlSmall<OutputSize>
```
You forgot to derive `Copy` and `Clone`.

newpavlov commented on groestl/src/lib.rs:
```diff
@@ -0,0 +1,128 @@
+}
+
+#[derive(Default)]
+pub struct GroestlBig<OutputSize>
```
`Copy` and `Clone`.

newpavlov reviewed (commented):
Several minor comments.

gsingh93 committed:
grostl: more review cleanup

gsingh93 committed:
grostl: added benchmarks

gsingh93 committed:
grostl: removed Digest implementation for Groestl

gsingh93 commented:
Ok, I made those changes and I added some benchmarks. Since I'm using the `standard_padding` function now, I'm going to assume everything is fine and I don't need to write any extra tests, so I think this should be good to go.

newpavlov commented:
Thank you for you work!

Unfortunately performance is a bit lower than I expected (slightly less than 1 MB/s), but I will leave it for future improvements, I'll add your links to the related issue.

I will publish this crate in the next few days.
